feat(sdk/dsl/compiler): dsl-compile --mode flag to turn on V2_COMPATIBLE, defaults to KF_PIPELINES_COMPILER_MODE env var. Fixes #5840

Bobgy opened the pull request:
**Description of your changes:**
Fixes https://github.com/kubeflow/pipelines/issues/5840

The env var is useful when a user has decided to use v2 compatible mode by default.

The user can set up the following in .bashrc:
```
export KF_PIPELINES_COMPILER_MODE=V2_COMPATIBLE
```

## Verification

I tested locally by installing `pip install -e sdk/python`.
And then dsl-compile
* without --mode flag
* with --mode flag
* without --mode flag, but set KF_PIPELINES_COMPILER_MODE env var

all of these cases worked as expected

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

chensun commented on sdk/python/kfp/compiler/main.py:
```diff
@@ -97,9 +101,19 @@ def main():
+  if args.mode:
+    if args.mode == 'V2_COMPATIBLE':
+      mode = dsl.PipelineExecutionMode.V2_COMPATIBLE
+    elif args.mode == 'V1_LEGACY' or args.mode == 'V1':
```
> `or args.mode == 'V1'`

For consistence's matter, we should either disallow this here, or allow it for the environment variable approach as well (which current raises a ValueError on this).

Bobgy replied:
That's a good catch! Let me fix this

Bobgy replied:
Updated, now we are consistent that 'V1' and 'V1_LEGACY' are both supported, while in user helper messages, we only show people 'V1', because 'V1' has not been deprecated.

chensun commented on sdk/python/kfp/compiler/compiler.py:
```diff
@@ -59,16 +61,29 @@ def my_pipeline(a: int = 1, b: str = "default value"):
         needed for tests or custom deployments right now.
     """
+    if mode is None:
+      mode_str = os.environ.get(KF_PIPELINES_COMPILER_MODE_ENV, 'V1_LEGACY')
```
Feels to me that allowing environmental variable is like opening a backdoor that breaks the interface. Env variable can be set or unset from anywhere and at anytime. The same code could behave differently on different machines or on the same machine but at different times. It could be a nightmare for debuggability. 
To me, the cons of using env variable outweighs the pros.

That being said, I'm okay if we want to use it here, just for this compatibility setting, given it's a transient setting which will go away when KFP v2 is ready, but not extending this pattern to other arguments.

Bobgy replied:
I agree that supporting env vars or not is a trade off between explicitness and convenience.

I have a different viewpoint on what matters in this scenario, remember that `dsl-compile` is a CLI tool. A CLI tool is used very often on different targets. It would be a lot more inconvenient if you use an argument all the time, but you have to type it each time. For the same reason, `kubectl`, `gcloud`, `go` etc CLI tools allow configuring some persistent configuration through env vars and/or config file, so that when you use them in a terminal. You'd only need to type the information related to your current command.

Allowing implicit configuration does make it harder to debug. So we need to spend more effort on this, e.g. `go env` prints out all your go environment variables, so that you can understand configuration for your current context. For our case, I've been using `env | grep KF` pretty happily to confirm my current setup. Maybe we can make that a formal command to make it easier to understand.

FYI, this is not the first introduction of env vars
https://github.com/kubeflow/pipelines/blob/8b9756247c68b681fc9a7c3ea1ce0dab5e22c847/sdk/python/kfp/_client.py#L84-L90
KFP client has supported configuration from env vars [since 0.1.25](https://github.com/kubeflow/pipelines/commit/6e5171452b923ba2da381d182a7d4e313d9f5321).

Another benefit of env vars, is that it can help decouple deployment specific configuration from pipeline/run specific configuration. e.g. if we document how users can configure env vars to let KFP client know how to talk to the API endpoint, then our samples do not need to always leave a placeholder for endpoint that people need to change before running. This can create an experience of always using the same command to run samples, see feature request https://github.com/kubeflow/pipelines/issues/5381

chensun replied:
Thanks for the detailed explanation. That makes sense. 
Given what you described, I would still prefer we have this env read logic in the CLI implementation (sdk/python/kfp/compiler/main.py), rather than inside the compiler (sdk/python/kfp/compiler/compiler.py). So that we can limit such pattern to cli/client use case only.

That said, it's still not a big deal for this particular argument.

Bobgy replied:
Good point, only adding env read logic in CLI implementation does feel better and enough for UX!

Bobgy replied:
Finished the implementation change

chensun reviewed (commented):
JFYI, we have some existing unit tests for the cli, for instance:
https://github.com/kubeflow/pipelines/blob/f5242666f6cdd5b9d9507699f9426d23625b529b/sdk/python/tests/compiler/compiler_tests.py#L314

But I won't worry about missing UTs for this particular change. It's up to you.

/lgtm

Bobgy commented on sdk/python/tests/compiler/compiler_tests.py:
```diff
@@ -383,6 +384,62 @@ def test_py_volume(self):
 
+  def test_py_compile_mode(self):
+    cases = [
+      {'mode': 'V2_COMPATIBLE', 'is_v2': True},
+      {'mode': 'V1', 'is_v2': False},
+      {'mode': 'V1_LEGACY', 'is_v2': False},
+      {'mode': None, 'is_v2': False},
+      {'mode': 'V2_COMPATIBLE', 'env': 'V1', 'is_v2': True},
+      {'mode': None, 'env': 'V1', 'is_v2': False},
+      {'mode': None, 'env': 'V2_COMPATIBLE', 'is_v2': True},
+      {'mode': None, 'env': 'V1_LEGACY', 'is_v2': False},
+      {'mode': 'INVALID', 'error': True},
+      {'mode': None, 'env': 'INVALID', 'error': True},
```
Hi @chensun, I'm adding a table driven test, because this feels like a right fit for full coverage.
How do you feel about this?

chensun replied:
A more standard way to do the same would be using parameterized tests: https://github.com/abseil/abseil-py/blob/master/absl/testing/parameterized.py

We also have examples in our existing UTs. For example:
https://github.com/kubeflow/pipelines/blob/2abc352ad8617afb1e299360a5b2a5f9da5772ae/sdk/python/kfp/dsl/type_utils_test.py#L139

Bobgy replied:
Thanks, updated to absl.testing.parameterized!
There's an obvious improvement on error message

chensun commented on sdk/python/kfp/compiler/main.py:
```diff
@@ -14,16 +14,14 @@
 from deprecated.sphinx import deprecated
 
-
+KF_PIPELINES_COMPILER_MODE_ENV = 'KF_PIPELINES_COMPILER_MODE'
```
nit: 2 empty lines between this and the following function definition.

Also maybe consider prefixing the varaible name with underscore: `_KF_PIPELINES_COMPILER_MODE_ENV = ...`? (assuming it's private, and referenced only within this file)

Bobgy replied:
updated.

I'm not sure why existing env vars did not have the prefix, but I think I agree with you it should probably be internal.

Bobgy commented:
Hi @chensun, thank you for the valuable review! learned a lot on python best practices.
I've addressed all your comments and it's ready for another round of review.

chensun commented:
/lgtm
/approve

Thank you!

chensun commented:
/retest

chensun commented:
/retest

james-jwu commented:
/cc @james-jwu

Bobgy commented:
/retest
